Engine/src/Engine/Rendering/Renderer.cpp

```diff
@@ -0,0 +1,255 @@
+	Renderer* Renderer::getInstance()
+	{
+		if (m_pinstance == nullptr)
+		{
+			m_pinstance = new Renderer();
+		}
+
+		return m_pinstance;
+	}
+
+	void Renderer::updateUIHeiraarchy(std::string tag, Entity entity)
+	{
+		m_UI.updateHierarchyPanel(tag, entity);
+	}
+
+#pragma region Scene Management & Rendering
+	//Initialize the Scene. Updates m_window's camera
+	void Renderer::initializeScene(Scene& scene)
+	{
+		//Initialize input system.
+		InputSystem::getInstance()->init(m_window.getWindow());
+
+		//initialize the window for UI
+		if (showUI)
+		{
+			if (!m_UI.initializeUI(m_window, scene))
+				GE_CORE_WARN("[Renderer] UI not initialized properly.");
+		}
+
+		//Create default shader.
+		loadShaders();
+
+		GLint prog = 0;
+		glGetIntegerv(GL_CURRENT_PROGRAM, &prog);
+
+		//Update m_window camera.
+		auto cameraView = scene.getEntities<CameraComponent>();
+		const auto camera = scene.m_registry.get<CameraComponent>(cameraView.back());
+		m_window.updateCamera(camera);
+	}
+
+	//For when the scene must be stopped, perform cleanup
+	void Renderer::stopScene(Scene& scene) 
+	{
+		glfwTerminate();
+		
+		glDeleteProgram(m_programId);
+	}
+
+	//From scene, we should grab all the entities within that scene object and render each one.
+	void Renderer::renderScene(const DeltaTime& dt, Scene& scene)
+	{
+		glClear(GL_COLOR_BUFFER_BIT);
+
+		//Bind the FBO for draw calls to render to.
+		m_window.bindFrameBuffer();
+
+		//Update window camera.
+		auto cameraView = scene.getEntities<CameraComponent>();
+		m_window.updateCamera(scene.m_registry.get<CameraComponent>(cameraView.back()));
+
+		//Render all entities with vertices, and associated components.
+		drawEntities(scene);
+
+		//Unbind the FBO.
+		m_window.unBindFrameBuffer();
+
+		//UI window
+		if (showUI) 
+		{
+			m_UI.renderUI(scene, m_window);
+		}
+		else
+		{
+			//Draw the FBO to the screen.
+			glBindTexture(GL_TEXTURE_2D, m_window.m_fboTextureID);
+			glDrawElements(GL_TRIANGLES, 6, GL_UNSIGNED_INT, nullptr);
+		}
+
+		//Swap the glfw buffers.
+		glfwSwapBuffers(m_window.getWindow());
+
+
+		//Check if the 'X' was hit on the window. If so, close the program
+		if (glfwWindowShouldClose(m_window.getWindow())) 
+		{
+			scene.m_closeScene = true;
+		}
+	}
+
+	//Draw all Entities within scene that contain a vertices component.
+	void Renderer::drawEntities(Scene& scene)
+	{
+		//Get entities that contain vertices component.
+		const auto renderables = scene.getEntities<const VerticesComponent>();
+
+		//Set the currentBound textures to 0 for this draw call.
+		int currentBoundTextures = 0;
+
+		//For each entitiy with a vertices component, render it.
+		for (auto [entity, vertices] : renderables.each())
+		{
+			//Set a default transform component and color if the object does not contain one.
+			TransformComponent transform{ glm::vec3 {0.f,0.f,0.f}, glm::vec3 {1.f,1.f,1.f} , glm::vec3 {0.f,0.f,0.f} };
+			glm::vec4 color{ 1.f,1.f,1.f,1.f };
```

Comment:
we probably want to use of the default constructor here instead of hardcoding these values